### PR TITLE
[JIT] Use type cache in erasing shape information

### PR DESCRIPTION
**Pull request by eellison**
`unshapedType` can be very slow on a graph with many modules and recursively contained classes, because each Type you have to recursively descend and map over. Speed it up with a type cache.

### CODE REVIEW

File: aten/src/ATen/core/jit_type.h [hunk 1]
```diff
@@ -1541,6 +1541,8 @@ TORCH_API std::ostream& operator<<(std::ostream& os, const Stride& s);
 // what is the type, ignoring extra size/shape information?
 // e.g. Tensor(2x3) -> Dynamic, and Tuple(Tensor(2x3),...) -> Tuple(Dynamic,...)
 
+// xxx: be careful with calls because this can be very slow. If calling this on a graph
+// use `EraseShapeInformation` in shape_analysis.h
 inline TypePtr unshapedType(const TypePtr& type) {
```
Comment (Krovatkin) [inline aten/src/ATen/core/jit_type.h:1546]: should we make `unshapedType` return a singleton?

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 2]
```diff
@@ -2172,29 +2172,62 @@ void PropagateInputShapes(const std::shared_ptr<Graph>& graph) {
...
+
+void EraseShapeInformation(const std::shared_ptr<Graph>& graph, TypeCache& cache) {
+  EraseShapeInformation(graph->block(), cache);
+}
+
 } // anonymous namespace
 
 void EraseShapeInformation(const std::shared_ptr<Graph>& graph) {
-  EraseShapeInformation(graph->block());
+  TypeCache cache;
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/shape_analysis.cpp:2229]: nitpick: maybe a small comment for better readability that this map maps shaped types to unshaped.

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 3]
```diff
@@ -2172,29 +2172,62 @@ void PropagateInputShapes(const std::shared_ptr<Graph>& graph) {
...
   for (Value* v : vals) {
-    v->setType(unshapedType(v->type()));
+    v->setType(getOrCreateUnshapedType(v->type(), cache));
   }
 }
 
-void EraseShapeInformation(Block* b) {
-  EraseShapeInformation(b->inputs());
-  EraseShapeInformation(b->outputs());
+void EraseShapeInformation(Block* b, TypeCache& cache) {
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/shape_analysis.cpp:2208]: nitpick: all this `cache` threading seems to beg for `EraseShapeInformation` class)

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 4]
```diff
@@ -2172,29 +2172,62 @@ void PropagateInputShapes(const std::shared_ptr<Graph>& graph) {
...
+  return type->withContained(unshaped_contained_types);
+}
+
+TypePtr getOrCreateUnshapedType(TypePtr type, TypeCache& cache) {
+  auto maybe_cached_type = cache.find(type);
+  if (maybe_cached_type != cache.end()) {
+    return maybe_cached_type->second;
+  }
+  auto unshaped_type = unshapedTypeImpl(type);
+  cache[type] = unshaped_type;
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/shape_analysis.cpp:2196]: nitpick: you could `insert` here, slightly more efficient and more canonical or the first-time insertions.

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: aten/src/ATen/core/jit_type.h (hunk 1 above)
Comment (eellison) [reply aten/src/ATen/core/jit_type.h:1546]: wdym ?

File: torch/csrc/jit/passes/shape_analysis.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/passes/shape_analysis.cpp:2229]: map maps maps on a sea shore

File: aten/src/ATen/core/jit_type.h (hunk 1 above)
Comment (eellison) [reply aten/src/ATen/core/jit_type.h:1546]: if you mean persistent caching, maybe, but that's a different PR. we would need to store weak ptrs and have all cache lock and such, i would wait till it's actually needed and/or fix callsites instead.

**Pull request closed without merging**